src/buffer/line_cache.rs

```diff
@@ -0,0 +1,115 @@
+use std::ops::Range;
+
+use crate::{document::ReadableDocument, simd::memchr2};
+
+/// Cache a line/offset pair every CACHE_EVERY lines to speed up line/offset calculations
+const CACHE_EVERY: usize = 1024 * 64;
+
+pub struct LineCache {
+    cache: Vec<(usize, usize)>, // (index, line)
+}
+
+impl LineCache {
+    pub fn new() -> Self {
+        Self { cache: vec![] }
+    }
+
+    pub fn from_document<T: ReadableDocument>(&mut self, document: &T) {
+        self.cache.clear();
+
+        let mut offset = 0;
+        let mut line = 0;
+        loop {
+            let text = document.read_forward(offset);
+            if text.is_empty() { return; }
+            
+            let mut off = 0;
+            loop {
+                off = memchr2(b'\n', b'\n', text, off);
+                if off == text.len() { break; }
+
+                if line % CACHE_EVERY == 0 {
+                    self.cache.push((offset+off, line));
+                }
+                line += 1;
+                off += 1;
+            }
+
+            offset += text.len();
+        }
+    }
+
+    /// Updates the cache after a deletion.
+    /// `range` is the deleted byte range, and `text` is the content that was deleted.
+    pub fn delete(&mut self, range: Range<usize>, text: &Vec<u8>) {
+        let mut newlines = 0;
+        for c in text {
+            if *c == b'\n' {
+                newlines += 1;
+            }
+        }
+
+        if self.cache.len() == 0 { return; }
+        
+        let len = text.len();
+        let mut i = self.cache.len() - 1;
+        loop {
+            let (ref mut off, ref mut line) = self.cache[i];
+            if *off > range.start {
+                if *off < range.end {
+                    self.cache.remove(i); // cache point is within the deleted range
```

Comment:
This is O(n^2) worst case if the user deletes the contents between the start of the file and the middle of it, if I'm reading this right. It would be better for us to first find the offset of the `range.start` item, then find the offset of the `range.end` item, and then delete all items in that range in a single `.drain()` call.